air-script/tests/memory/memory.air

```diff
@@ -0,0 +1,71 @@
+def MemoryAir
+
+trace_columns:
+    main: [s[2], ctx, addr, clk, v[4], d[2], t, dummy]
+
+public_inputs:
+    stack_inputs: [16]
+
+random_values:
+    alphas: [9]
+
+boundary_constraints:
+    # This is a dummy trace column to satisfy requirement of at least one boundary constraint.
+    enf dummy.first = 0
+
+integrity_constraints:
+    let delta_c = ctx' - ctx
+    let delta_a = addr' - addr
+    let delta_i = clk' - clk - 1
+
+    # n_0 = 1 when context changes and 0 otherwise.
+    let n_0 = delta_c * t'
+    # if context remains the same, n_1 = 1 when address changes and 0 otherwise.
+    let n_1 = delta_a * t'
+
+    # Enforce that n_0 must be binary.
+    enf n_0^2 - n_0 = 0
+    # Enforce that when context changes, n_0 = 1
+    enf (1 - n_0) * delta_c = 0
+    # Enforce that n_1 must be binary. An additional multiplier ensures that the check of n_1 
+    # occurs only if the context does not change.
+    enf (1 - n_0) * (n_1^2 - n_1) = 0
+    # Enforce that if context remains the same, n_1 = 1 when address changes and 0 otherwise.
+    enf (1 - n_0) * (1 - n_1) * delta_a = 0
+
+    # Enforce that values in the selectior columns must be binary.
+    # s[0] is set to 0 for write operations and to 1 for read operations
+    enf s[0]^2 - s[0] = 0
+    enf s[1]^2 - s[1] = 0
+
+    # Enforce that s[1] = 1 when the operation is a read and `ctx` and `addr` columns are both 
+    # unchanged.
+    (1 - n_0) * (1 - n_1) * s[0]' * (1 - s[1]') = 0
+    # Enforce that when either the context changed, the address changed, or the operation is a 
+    # write, then s[1] = 0
+    (n_0 + (1 - n_0) * n_1 + (1 - s[0]')) * s[1]' = 0
+
+    let d_next_aggr = 2^16 * d[1]' + d[0]'
+
+    # Enforce that values of context (`ctx`), address (`addr`), and clock cycle (`clk`) grow 
+    # monotonically
+    enf (n_0 * delta_c + (1 - n_0) * (n_1 * delta_a + (1 - n_1) * delta_i)) - d_next_aggr = 0
+
+    # TODO: perform range checks for values in columns d[0] and d[1]
+    # How can I call a range check for some columns?
+
+    # Enforce that values at a given memory address are always initialized to 0.
+    s[0] * (1 - s[1]) * v[0] = 0
+    s[0] * (1 - s[1]) * v[1] = 0
+    s[0] * (1 - s[1]) * v[2] = 0
+    s[0] * (1 - s[1]) * v[3] = 0
+
+    # Enforce that for the same context/address combination, the v[i] columns of the current row 
+    # are equal to the corresponding u[i] columns of the next row, for i from 0 to 3
+    enf s[1] * (v[0]' - v[0]) = 0
+    enf s[1] * (v[1]' - v[1]) = 0
+    enf s[1] * (v[2]' - v[2]) = 0
+    enf s[1] * (v[3]' - v[3]) = 0
```

Comment:
Question: Should we wait to merge this and first merge constraint comprehension PR first? I think we should be able to merge it sometime next week and we'll be able to write these constraints in a much more concise way once that PR is merged.
cc: @grjte

Reply:
Or maybe we could mention it in comments that these could probably be written as:
```
fn change(v: scalar) -> scalar:
    return v' - v
enf s[1] * change(v[i]) = 0 for i in 0..4
```
once pure functions and constraint comprehension are implemented.

Wdyt?

Reply:
I agree, it looks much more concise. I think it's worth waiting if it doesn't create additional problems or delays.

Reply:
I don't think we should wait. I think we should just move where we put this file. Instead of putting these constraints into the `tests`, we should create a new directory and put all our constraint definitions for the miden vm there. That way the VM constraints can be defined before the syntax is supported. If we don't do that, I don't think we'll get everything done in time.

Reply:
I don't think we should wait. I think we should just move where we put this file. Instead of putting these constraints into the `tests`, we should create a new directory and put all our constraint definitions for the miden vm there. That way the VM constraints can be defined before the syntax is supported. If we don't do that, I don't think we'll get everything done in time.